src/java.base/share/classes/jdk/internal/lang/stable/StableValueImpl.java

```diff
@@ -0,0 +1,191 @@
+/**
+ * The implementation of StableValue.
+ *
+ * @implNote This implementation can be used early in the boot sequence as it does not
+ *           rely on reflection, MethodHandles, Streams etc.
+ *
+ * @param <T> type of the underlying data
+ */
+public final class StableValueImpl<T> implements StableValue<T> {
+
+    // Unsafe allows StableValue to be used early in the boot sequence
+    static final Unsafe UNSAFE = Unsafe.getUnsafe();
+
+    // Unsafe offsets for direct field access
+    private static final long UNDERLYING_DATA_OFFSET =
+            UNSAFE.objectFieldOffset(StableValueImpl.class, "value");
+
+    // Generally, fields annotated with `@Stable` are accessed by the JVM using special
+    // memory semantics rules (see `parse.hpp` and `parse(1|2|3).cpp`).
+    //
+    // This field is used directly and via Unsafe using explicit memory semantics.
+    //
+    // | Value          |  Meaning      |
+    // | -------------- |  ------------ |
+    // | null           |  Unset        |
+    // | nullSentinel() |  Set(null)    |
+    // | other          |  Set(other)   |
+    //
+    @Stable
+    private volatile Object value;
+
+    // Only allow creation via the factory `StableValueImpl::newInstance`
+    private StableValueImpl() {}
+
+    @ForceInline
+    @Override
+    public boolean trySet(T value) {
+        if (this.value != null) {
+            return false;
+        }
+        // Mutual exclusion is required here as `computeIfUnset` might also
+        // attempt to modify the `wrappedValue`
+        synchronized (this) {
+            return wrapAndCas(value);
+        }
+    }
+
+    @ForceInline
+    @Override
+    public void setOrThrow(T value) {
+        if (!trySet(value)) {
+            throw new IllegalStateException("Cannot set the underlying data to " + value +
+                    " because the underlying data is already set: " + this);
+        }
+    }
+
+    @ForceInline
+    @Override
+    public T orElseThrow() {
+        final Object t = value;
+        if (t == null) {
+            throw new NoSuchElementException("No underlying data set");
+        }
+        return unwrap(t);
+    }
+
+    @ForceInline
+    @Override
+    public T orElse(T other) {
+        final Object t = value;
+        return (t == null) ? other : unwrap(t);
+    }
+
+    @ForceInline
+    @Override
+    public boolean isSet() {
+        return value != null;
+    }
+
+    @ForceInline
+    @Override
+    public T orElseSet(Supplier<? extends T> supplier) {
+        final Object t = value;
+        return (t == null) ? computeIfUnsetSlowPath(supplier) : unwrap(t);
+    }
+
+    @DontInline
+    private synchronized T computeIfUnsetSlowPath(Supplier<? extends T> supplier) {
+        final Object t = value;
+        if (t == null) {
+            final T newValue = supplier.get();
+            // The mutex is reentrant so we need to check if the value was actually set.
+            return wrapAndCas(newValue) ? newValue : orElseThrow();
```

Comment:
Reentrancy into here seems really buggy, I would endorse disallowing it instead. In that case, a `ReentrantLock` seems better than the native monitor as we can cheaply check `lock.isHeldByCurrentThread()`

Reply:
StableValueImpl was carefully designed to minimize memory footprint. Adding a lock would inflate memory usage substantially.

Reply:
There is also `Thread.holdsLock()`.